lib/fhir/common.js

```diff
@@ -119,6 +110,184 @@ module.exports = exports = (mongo) => {
+      if (versionId) {
+        db
+          .collection(`${resourceType}_history`)
+          .findOne({id: resourceId, 'meta.versionId': versionId}, callback)
+      } else {
+        db.collection(resourceType).findOne({id: resourceId}, callback)
+      }
+    })
+  }
+
+  function resolveNestedProperties (resultList, includeReference, callback) {
+    let promises = []
+
+    // resolve references in nested objects recursively
+    const evaluateProperties = (property, nextProperty) => {
+      if (property == null) {
+        return callback(new Error('Undefined resource property.'))
+      }
+
+      if (Array.isArray(property)) {
+        for (var n = 0; n < property.length; n++) {
+          if (nextProperty[0] && property[n][nextProperty[0]] != null) {
+            evaluateProperties(property[n][nextProperty[0]], nextProperty.slice(1))
+          } else if (property[n].reference) {
+            promises.push(new Promise((resolve, reject) => {
+              resolveReference(property[n].reference, (err, data) => {
+                if (err) {
+                  reject(err)
+                } else {
+                  resolve(data)
+                }
+              })
+            }))
+          } else {
+            return callback(new Error(`Invalid resource reference: "${JSON.stringify(property[n])}"`))
+          }
+        }
+      } else {
+        if (nextProperty[0] && property[nextProperty[0]] != null) {
+          evaluateProperties(property[nextProperty[0]], nextProperty.slice(1))
+        } else if (property.reference) {
+          promises.push(new Promise((resolve, reject) => {
+            resolveReference(property.reference, (err, data) => {
+              if (err) {
+                reject(err)
+              } else {
+                resolve(data)
+              }
+            })
+          }))
+        } else {
+          return callback(`Invalid resource reference: "${JSON.stringify(property)}"`)
+        }
+      }
+    }
+
+    evaluateProperties(resultList[includeReference[0]], includeReference.slice(1))
+
+    // resolve promises
+    Promise.all(promises)
+      .then((res) => {
+        callback(null, res)
+      }).catch((err) => {
+        callback(err)
+      })
+  }
+
+  const parseSearchExpression = (exp, callback) => {
+    const itemList = exp.split(':')
+    const itemResource = itemList[0]
+    const itemReference = itemList[1].split('.')
+
+    callback(itemResource, itemReference)
+  }
+
+  const resolveIncludeParameters = (item, results) => {
+    return new Promise((resolve, reject) => {
+      const itemList = item.split('.')
+      const itemResource = itemList[0]
+      const itemReference = itemList.slice(1)
+
+      results.forEach((item) => {
+        if (item['resourceType'] === itemResource) {
+          resolveNestedProperties(item, itemReference, (err, data) => {
+            if (err) {
+              reject(err)
+            } else {
+              resolve(data)
```

Comment:
We shouldn't be calling resolve in a loop. This could resolve or reject the same promise multiple times.